crates/ress/provider/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+        }
+
+        // NOTE: there might be a race condition where target ancestor hash gets evicted from the
+        // database.
+        let witness_state_provider = self.provider.state_by_block_hash(ancestor_hash)?;
+        let mut trie_input = TrieInput::default();
+        for block in executed_ancestors.into_iter().rev() {
+            trie_input.append_cached_ref(&block.trie, &block.hashed_state);
+        }
+        let mut hashed_state = db.into_state();
+        hashed_state.extend(record.hashed_state);
+
+        // Gather the state witness.
+        let witness = if hashed_state.is_empty() {
+            // If no state was accessed, at least the root node must be present.
+            let multiproof = witness_state_provider.multiproof(
+                trie_input,
+                MultiProofTargets::from_iter([(B256::ZERO, Default::default())]),
+            )?;
+            let mut witness = Vec::new();
+            if let Some(root_node) =
+                multiproof.account_subtree.into_inner().remove(&Nibbles::default())
+            {
+                witness.push(root_node);
+            }
+            witness
+        } else {
+            witness_state_provider.witness(trie_input, hashed_state)?
+        };
+
+        // Insert witness into the cache.
+        self.witness_cache.lock().insert(block_hash, Arc::new(witness.clone()));
+
+        Ok(witness)
+    }
+}
+
+impl<N, E> RessProtocolProvider for RethRessProtocolProvider<N, E>
+where
+    N: ProviderNodeTypes<Primitives = EthPrimitives>,
+    E: BlockExecutorProvider<Primitives = N::Primitives> + Clone,
+{
+    fn header(&self, block_hash: B256) -> ProviderResult<Option<Header>> {
+        trace!(target: "reth::ress_provider", %block_hash, "Serving header");
+        Ok(self.block_by_hash(block_hash)?.map(|b| b.header().clone()))
+    }
+
+    fn block_body(&self, block_hash: B256) -> ProviderResult<Option<BlockBody>> {
+        trace!(target: "reth::ress_provider", %block_hash, "Serving block body");
+        Ok(self.block_by_hash(block_hash)?.map(|b| b.body().clone()))
+    }
+
+    fn bytecode(&self, code_hash: B256) -> ProviderResult<Option<Bytes>> {
+        trace!(target: "reth::ress_provider", %code_hash, "Serving bytecode");
+        let maybe_bytecode = 'bytecode: {
+            if let Some(bytecode) = self.pending_state.find_bytecode(code_hash) {
+                break 'bytecode Some(bytecode);
+            }
+
+            self.provider.latest()?.bytecode_by_hash(&code_hash)?
+        };
+
+        Ok(maybe_bytecode.map(|bytecode| bytecode.original_bytes()))
+    }
+
+    async fn witness(&self, block_hash: B256) -> ProviderResult<Vec<Bytes>> {
+        trace!(target: "reth::ress_provider", %block_hash, "Serving witness");
+        let started_at = Instant::now();
+        let this = self.clone();
+        match self.witness_task_pool.spawn(move || this.generate_witness(block_hash)).await {
```

Comment:
I think I'd rather use `Box<dyn TaskSpawner>` and use .spawn_blocking here for all of the these workloads

Reply:
how's that resolved?

Reply:
overlooked, resolved now